File: dev/testSk3/HelloSk.Core/GoogleAdsPlugin.fs
namespace HelloSk.Core

open System
open System.Collections.Generic
open System.ComponentModel
open System.Net.Http
open System.Text
open System.Text.Json
open Microsoft.SemanticKernel

[<AutoOpen>]
module private GoogleAdsPluginImpl =
    let tokenUrl = "https://www.googleapis.com/oauth2/v3/token"
    let searchVersion = "v21"
    let defaultGaql = "SELECT campaign.id, campaign.name FROM campaign LIMIT 10"

    let getEnv () : Result<Map<string, string>, string> =
        Shared.getEnv "GOOGLE_ADS_PLUGIN_ENV"
        |> Option.filter (String.IsNullOrWhiteSpace >> not)
        |> function
            | Some p -> Shared.loadEnvFile p
            | None -> Shared.loadEnvFromCandidates [ "googleapi/.env"; ".env" ]

    let requiredKeys = [ "client_id"; "client_secret"; "refresh_token"; "developer_token"; "login-customer-id" ]

    let getAccessToken (client: HttpClient) (env: Map<string, string>) : Result<string, string> =
        let form = [
            KeyValuePair<string, string>("grant_type", "refresh_token")
            KeyValuePair<string, string>("client_id", env |> Map.find "client_id")
            KeyValuePair<string, string>("client_secret", env |> Map.find "client_secret")
            KeyValuePair<string, string>("refresh_token", env |> Map.find "refresh_token")
        ]
        let content = new FormUrlEncodedContent(form)
        try
            let resp = client.PostAsync(tokenUrl, content).GetAwaiter().GetResult()
            let body = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult()
            if not resp.IsSuccessStatusCode then
                Error(sprintf "取得 token 失敗 (%d): %s" (int resp.StatusCode) body)
            else
                use doc = JsonDocument.Parse(body)
                let root = doc.RootElement
                let mutable tokEl = Unchecked.defaultof<JsonElement>
                if root.TryGetProperty("access_token", &tokEl) then
                    Ok(tokEl.GetString() |> Option.ofObj |> Option.defaultValue "")
                else
                    Error("access_token not in response: " + body)
        with ex -> Error(ex.Message)

    let search (client: HttpClient) (env: Map<string, string>) (accessToken: string) (customerId: string) (query: string) : Result<string, string> =
        let url = sprintf "https://googleads.googleapis.com/%s/customers/%s/googleAds:search" searchVersion customerId
        let body = JsonSerializer.Serialize({| query = query |})
        use req = new HttpRequestMessage(HttpMethod.Post, url)
        req.Content <- new StringContent(body, Encoding.UTF8, "application/json")
        req.Headers.TryAddWithoutValidation("Authorization", "Bearer " + accessToken) |> ignore
        req.Headers.TryAddWithoutValidation("developer-token", env |> Map.find "developer_token") |> ignore
        req.Headers.TryAddWithoutValidation("login-customer-id", env |> Map.find "login-customer-id") |> ignore
        try
            let resp = client.SendAsync(req).GetAwaiter().GetResult()
            let s = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult()
            if not resp.IsSuccessStatusCode then
                Error(sprintf "查詢失敗 (%d): %s" (int resp.StatusCode) s)
            else
                Ok s
        with ex -> Error(ex.Message)

/// 查詢 Google Ads API 的 Plugin，使用 GAQL。需在 .env 設定 client_id, client_secret, refresh_token, developer_token, login-customer-id。
/// .env 可放在專案根目錄或 googleapi/ 目錄；或設定環境變數 GOOGLE_ADS_PLUGIN_ENV 指定檔案路徑。
type GoogleAdsPlugin() =

    [<KernelFunction("query_ads")>]
    [<Description("用 GAQL 查詢 Google Ads。傳入 GAQL 查詢字串（例如查 campaign、ad_group、keyword_view 等）；可選傳入 customer_id 覆寫查詢的客戶 ID。若不傳 gaql 則使用預設查詢（campaign 列表）。")>]
    member _.QueryAds(
        [<Description("GAQL 查詢字串，例如: SELECT campaign.id, campaign.name FROM campaign LIMIT 5")>] gaql: string,
        [<Description("選填。要查詢的 Google Ads 客戶 ID，不傳則用 .env 的 customer-id 或 login-customer-id")>] customerId: string
    ) : string =
        let query =
            (gaql |> Option.ofObj |> Option.defaultValue "").Replace("\n", " ").Trim()
            |> fun s -> if String.IsNullOrEmpty s then defaultGaql else s

        match getEnv () with
        | Error e ->
            JsonSerializer.Serialize({| error = "缺少 .env: " + e |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
        | Ok env ->
            let missing = requiredKeys |> List.filter (fun k -> Option.isNone (Map.tryFind k env))
            if missing.Length > 0 then
                JsonSerializer.Serialize({| error = sprintf "缺少 .env 變數: %s" (String.concat ", " missing) |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
            else
                let cid =
                    (customerId |> Option.ofObj |> Option.defaultValue "").Trim()
                    |> fun s ->
                        if String.IsNullOrEmpty s then
                            env |> Map.tryFind "customer-id" |> Option.orElse (env |> Map.tryFind "login-customer-id") |> Option.defaultValue ""
                        else
                            s
                if String.IsNullOrEmpty cid then
                    JsonSerializer.Serialize({| error = "未設定 customer_id 且 .env 無 customer-id / login-customer-id" |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
                else
                    use client = new HttpClient()
                    client.Timeout <- TimeSpan.FromSeconds(60.0)
                    match getAccessToken client env with
                    | Error msg ->
                        JsonSerializer.Serialize({| error = msg |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
                    | Ok accessToken ->
                        match search client env accessToken cid query with
                        | Error msg ->
                            JsonSerializer.Serialize({| error = msg; query = query |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
                        | Ok body ->
                            use doc = JsonDocument.Parse(body)
                            let root = doc.RootElement
                            let mutable errEl = Unchecked.defaultof<JsonElement>
                            if root.TryGetProperty("error", &errEl) then
                                JsonSerializer.Serialize({| error = errEl.GetRawText(); query = query |}, JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping))
                            else
                                let results = ResizeArray<string>()
                                let mutable resultsEl = Unchecked.defaultof<JsonElement>
                                if root.TryGetProperty("results", &resultsEl) && resultsEl.ValueKind = JsonValueKind.Array then
                                    results.Add(sprintf "共 %d 筆" (resultsEl.GetArrayLength()))
                                    let arr = resultsEl
                                    for i = 0 to arr.GetArrayLength() - 1 do
                                        results.Add(sprintf "  [%d] %s" (i + 1) (arr.[i].GetRawText()))
                                else
                                    results.Add("共 0 筆")
                                results.Add("")
                                results.Add("完整回應:")
                                let jsonOpts = JsonSerializerOptions(WriteIndented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping)
                                results.Add(JsonSerializer.Serialize(root, jsonOpts))
                                String.concat "\n" results

File: dev/testSK2/google_ads_plugin.py
# -*- coding: utf-8 -*-
"""
Google Ads API Plugin（參照 googleapi SKILL）

提供 query_ads：以 GAQL 查詢 Google Ads，流程與 query-ads.js / SKILL.md 一致：
1. 從 .env 讀取 client_id, client_secret, refresh_token, developer_token, login-customer-id（可選 customer-id）
2. 用 refresh_token 取得 access_token
3. 呼叫 googleAds:search 並回傳 results 摘要

.env 可放在專案根目錄或 googleapi/ 目錄。
"""
import json
import os
import urllib.error
import urllib.parse
import urllib.request
from typing import Annotated

from semantic_kernel.functions import kernel_function

TOKEN_URL = "https://www.googleapis.com/oauth2/v3/token"
SEARCH_VERSION = "v21"


def _load_env() -> dict[str, str]:
    """從 googleapi/.env 或專案根 .env 讀取（與 SKILL 同目錄優先），回傳 key=value 字典。"""
    root = os.path.dirname(os.path.abspath(__file__))
    env = {}
    for base in (os.path.join(root, "googleapi"), root):
        path = os.path.join(base, ".env")
        if not os.path.isfile(path):
            continue
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                idx = line.find("=")
                if idx > 0:
                    env[line[:idx].strip()] = line[idx + 1 :].strip()
        break
    return env


def _get_access_token(env: dict[str, str]) -> str:
    """用 refresh_token 取得 access_token。"""
    body = "&".join([
        "grant_type=refresh_token",
        f"client_id={urllib.parse.quote(env['client_id'], safe='')}",
        f"client_secret={urllib.parse.quote(env['client_secret'], safe='')}",
        f"refresh_token={urllib.parse.quote(env['refresh_token'], safe='')}",
    ])
    req = urllib.request.Request(
        TOKEN_URL,
        data=body.encode("utf-8"),
        method="POST",
        headers={"Content-Type": "application/x-www-form-urlencoded"},
    )
    with urllib.request.urlopen(req, timeout=30) as res:
        data = json.loads(res.read().decode())
    token = data.get("access_token")
    if not token:
        raise RuntimeError(f"取得 token 失敗: {data}")
    return token


def _search(env: dict[str, str], access_token: str, customer_id: str, query: str) -> dict:
    """呼叫 googleAds:search，回傳完整 JSON。"""
    url = f"https://googleads.googleapis.com/{SEARCH_VERSION}/customers/{customer_id}/googleAds:search"
    body = json.dumps({"query": query}).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {access_token}",
            "developer-token": env["developer_token"],
            "login-customer-id": env["login-customer-id"],
        },
    )
    with urllib.request.urlopen(req, timeout=60) as res:
        return json.loads(res.read().decode())


class GoogleAdsPlugin:
    """查詢 Google Ads API 的 Plugin，使用 GAQL。需在 .env 設定 OAuth 與 developer 變數。"""

    @kernel_function(
        name="query_ads",
        description="用 GAQL 查詢 Google Ads。傳入 GAQL 查詢字串（例如查 campaign、ad_group、keyword_view 等）；可選傳入 customer_id 覆寫查詢的客戶 ID。若不傳 gaql 則使用預設查詢（campaign 列表）。",
    )
    def query_ads(
        self,
        gaql: Annotated[str, "GAQL 查詢字串，例如: SELECT campaign.id, campaign.name FROM campaign LIMIT 5"] = "",
        customer_id: Annotated[str, "選填。要查詢的 Google Ads 客戶 ID，不傳則用 .env 的 customer-id 或 login-customer-id"] = "",
    ) -> str:
        query = (gaql or "").strip().replace("\n", " ").strip()
        print("query:")
        print(query)
        env = _load_env()
        required = ["client_id", "client_secret", "refresh_token", "developer_token", "login-customer-id"]
        missing = [k for k in required if not env.get(k)]
        if missing:
            return json.dumps({"error": f"缺少 .env 變數: {', '.join(missing)}"}, ensure_ascii=False)

        cid = (customer_id or "").strip() or env.get("customer-id") or env.get("login-customer-id")
        if not cid:
            return json.dumps({"error": "未設定 customer_id 且 .env 無 customer-id / login-customer-id"}, ensure_ascii=False)

        try:
            access_token = _get_access_token(env)
        except urllib.error.HTTPError as e:
            body = e.read().decode() if e.fp else str(e)
            return json.dumps({"error": f"取得 token 失敗 ({e.code}): {body}"}, ensure_ascii=False)
        except Exception as e:
            return json.dumps({"error": f"取得 token 失敗: {e!s}"}, ensure_ascii=False)

        try:
            search_res = _search(env, access_token, cid, query)
        except urllib.error.HTTPError as e:
            body = e.read().decode() if e.fp else str(e)
            return json.dumps({"error": f"查詢失敗 ({e.code}): {body}", "query": query}, ensure_ascii=False)
        except Exception as e:
            return json.dumps({"error": f"查詢失敗: {e!s}", "query": query}, ensure_ascii=False)

        if search_res.get("error"):
            return json.dumps({"error": search_res["error"], "query": query}, ensure_ascii=False)

        results = search_res.get("results") or []
        lines = [f"共 {len(results)} 筆"]
        for i, row in enumerate(results, 1):
            lines.append(f"  [{i}] {json.dumps(row, ensure_ascii=False)}")
        lines.append("\n完整回應:")
        lines.append(json.dumps(search_res, ensure_ascii=False, indent=2))
        print("search_res:")
        print(search_res)
        return "\n".join(lines)
